Forms: Consolidate TypeScript type definitions

edanzer opened the pull request:
## Proposed changes:

**Context:** We've started adding TypeScript to the forms package. I noticed we have two types.ts files in src/dashboard, and some of the types there will be reused outside of src/dashboard in src/blocks in soon-to-come PRs. 

**Changes:** 
- Move types from `dashboard/types.ts` and `dashboard/integrations/types.ts` to a single `src/types.ts` that can hold canonical shared types for the forms package. 
- Add detailed docs for those types
- Update any relevant imports

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
None. 

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No. 

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

This is a small refactor, so functionality should remain unchanged. Just ensure no build errors and smoke test related screens, including the Forms > About and Forms > Integrations screen (to see Forms > Integrations, you'll need to enable a feature flag with `add_filter( 'jetpack_forms_enable_integrations_tab', '__return_true' );`)

edanzer committed:
Consolidate forms TS types

edanzer committed:
Update type imports in two other files

edanzer committed:
changelog

edanzer committed:
Add docs for types

manzoorwanijk commented on projects/packages/forms/src/types/types.ts:
```diff
@@ -0,0 +1,75 @@
+ * @property title       The display name of the pattern.
+ * @property recommended Whether this pattern is recommended for most users.
+ * @property code        The unique code identifier for the pattern.
+ * @property description A short description of the pattern's purpose.
```
Instead of documenting here, it's better to document them inline to have IDEs pick up the description.

This is how the description looks in VS Code when you hover over the property.
<img width="763" alt="Screenshot 2025-06-05 at 7 47 41 PM" src="https://github.com/user-attachments/assets/95b12912-eca3-4e9e-adbe-f3a313d5af15" />

If you instead do it like this

```ts
/**
 * Represents a reusable form pattern for the Jetpack Forms dashboard.
 */
export type Pattern = {
	/**
	 * The unique code identifier for the pattern.
	 */
	code: string;
};
```
Here is how the IDEs will show it.

<img width="773" alt="Screenshot 2025-06-05 at 7 48 39 PM" src="https://github.com/user-attachments/assets/529d7338-00fc-4724-a365-f8636398f5ee" />

You can add all the doc block tags there, like `@default`, `@see`, `@link` etc., to make it even more useful.

edanzer replied:
Great suggestion @manzoorwanijk. Updated.

manzoorwanijk commented on projects/packages/forms/src/types/types.ts:
```diff
@@ -0,0 +1,75 @@
+ * Extends the Window interface to include Jetpack Forms block defaults.
+ * @property jpFormsBlocks Optional Jetpack Forms block defaults on the window object.
+ */
+export interface JPFormsBlocksWindow extends Window {
+	jpFormsBlocks?: {
+		defaults?: JPFormsBlocksDefaults;
+	};
+}
```
A better way to declare properties on global objects is to use `declare` keyword like this

https://github.com/Automattic/jetpack/blob/466baf982e6bf74c17a479f1c2c8b0ca4613b632/projects/packages/external-media/src/shared/media-service/index.ts#L90-L97

manzoorwanijk replied:
So, you can replace this with 

```ts
declare global {
	interface Window {
		jpFormsBlocks?: {
			defaults?: JPFormsBlocksDefaults;
		};
	}
}
```
which means that you will not need to do the assertion of the `window` object here

https://github.com/Automattic/jetpack/blob/466baf982e6bf74c17a479f1c2c8b0ca4613b632/projects/packages/forms/src/dashboard/integrations/google-sheets-card.tsx#L9

edanzer replied:
Also great suggestion, and also updated.

edanzer committed:
Fix import path after rebase

edanzer committed:
Fix type import after rebase

edanzer committed:
Document each type property separately

edanzer committed:
Improve handling of JPFormsBlocksDefaults type

edanzer committed:
Update types.ts to index.ts

edanzer commented:
I've made the following changes in response to feedback. cc @manzoorwanijk @dhasilva
- Updated types/type.ts to types/index.ts for shorter imports
- Updated documentation in types/index.ts so each property is separately documented
- Updated how we extend the global window object for JPFormsBlocksWindow

edanzer committed:
Update imports in newer TS files

dhasilva reviewed (approved):
LGTM :rocket:
